typescript/sdk/src/providers/MultiProvider.ts

```diff
@@ -1,139 +1,272 @@
-import { Signer, ethers } from 'ethers';
-
-import { ChainMap, ChainName, IChainConnection, Remotes } from '../types';
-import { MultiGeneric } from '../utils/MultiGeneric';
-import { objMap, pick } from '../utils/objects';
-
-import { ChainConnection } from './ChainConnection';
-
-export class MultiProvider<
-  Chain extends ChainName = ChainName,
-> extends MultiGeneric<Chain, ChainConnection> {
-  constructor(chainConnectionConfigs: ChainMap<Chain, IChainConnection>) {
-    super(
-      objMap(
-        chainConnectionConfigs,
-        (_, connection) => new ChainConnection(connection),
-      ),
-    );
+import { Debugger, debug } from 'debug';
+import {
+  BigNumber,
+  ContractReceipt,
+  ContractTransaction,
+  PopulatedTransaction,
+  Signer,
+  providers,
+} from 'ethers';
+
+import { types } from '@hyperlane-xyz/utils';
+
+import { ChainMetadata, chainMetadata } from '../consts/chainMetadata';
+import { CoreChainName, TestChains } from '../consts/chains';
+import { ChainMap, ChainName } from '../types';
+import { pick } from '../utils/objects';
+
+type Provider = providers.Provider;
+
+interface Params {
+  metadata?: ChainMap<ChainMetadata>;
+  providers?: ChainMap<Provider>;
+  signers?: ChainMap<Signer>;
+}
+
+export class MultiProvider {
+  public readonly metadata: ChainMap<ChainMetadata>;
+  private readonly providers: ChainMap<Provider>;
+  private readonly signers: ChainMap<Signer>;
+  private readonly logger: Debugger;
+
+  constructor({ metadata, providers, signers }: Params = {}) {
+    this.metadata = metadata ?? chainMetadata;
+    this.providers = providers ?? {};
+    this.signers = signers ?? {};
+    this.logger = debug('hyperlane:MultiProvider');
   }
 
   /**
-   * Get chainConnection for a chain
-   * @throws if chain is invalid or has not been set
+   * Get the metadata for a given chain name or id
+   * @throws if chain's metadata has not been set
    */
-  getChainConnection(chain: Chain): ChainConnection {
-    return this.get(chain);
+  tryGetChainMetadata(chainNameOrId: ChainName | number): ChainMetadata | null {
+    let chainMetadata: ChainMetadata | undefined;
+    if (typeof chainNameOrId === 'string') {
+      chainMetadata = this.metadata[chainNameOrId];
+    } else {
+      chainMetadata = Object.values(this.metadata).find(
+        (m) => m.id === chainNameOrId,
```

Comment:
As per the above comment, `id` is assumed to be the domain ID, not the chain ID?